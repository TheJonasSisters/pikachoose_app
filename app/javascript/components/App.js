import React, { Component } from "react"
import { BrowserRouter as Router, Route } from "react-router-dom"
import Header from "./components/Header"
import HistoryIndex from "./pages/HistoryIndex.js"
import HistoryShow from "./pages/HistoryShow.js"
import About from "./pages/About.js"
import LandingPage from "./pages/LandingPage.js"
import FavoriteIndex from "./pages/FavoriteIndex.js"
import FavoriteShow from "./pages/FavoriteShow.js"

class App extends Component {
  constructor() {
    super()
    this.state = {
      // filmList: [],
      histories: [],
      favorites: [],
      editable: null
    }
  }

  componentDidMount() {
    // this.getFilms()
    this.getHistories()
    this.getFavorites()
  }

  // generateRandomDecisionIndex = () => {
  //   const { filmList } = this.state
  //   // generate random number between 0 and filmList length
  //   let randomIndex = Math.floor(Math.random() * Math.floor(filmList.length))
  //   // retrieve random film from filmList
  //   let decision = filmList[randomIndex]
  //   // store the single record above to filmDecision
  //   return decision
  // }
  //
  // getFilms = () => {
  //   const apiKey = process.env.REACT_APP_KEY
  //
  //   fetch(`https://api.themoviedb.org/3/movie/top_rated?api_key=${apiKey}&language=en-US&page=1`)
  //   .then((response) => {
  //     if (response.status === 200) {
  //       return response.json()
  //     }
  //   })
  //   .then((films) => {
  //     films = films.results
  //     this.setState({filmList: films})
  //     console.log("Entire Film List:", this.state.filmList)
  //
  //     // results come back empty if the state is deconstructed
  //     // const { filmList, filmDecision } = this.state
  //     // let decision = this.generateRandomDecisionIndex()
  //     // this.setState({filmDecision: decision})
  //     // console.log("Film Decision:", this.state.filmDecision)
  //   })
  // }

  getHistories = () => {
    fetch("/histories")
    .then((response) => {
      if (response.status === 200) {
        return response.json()
      }
    })
    .then((histories) => {
      this.setState({histories: histories})
    })
  }

  getFavorites = () => {
    fetch("/favorites")
    .then((response) => {
      if (response.status === 200) {
        return response.json()
      }
    })
    .then((favorites) => {
      this.setState({favorites: favorites})
    })
  }

  addFavorite = (film) => {
    console.log("addFavorite(film):",film)
    let newFavorite = {
      film_id: film.id,
      title: film.title,
      overview: film.overview,
      vote_average: film.vote_average,
      release_date: film.release_date,
      comment: "No comments yet"
    }

    // fetch method gets specific film with the id in our back-end and UPDATES it
    fetch("/favorites",
    {
      method: 'POST',
      body: JSON.stringify(newFavorite),
      headers: {
        'Content-Type': 'application/json'
      }
    })

    console.log("newFavorite:", JSON.stringify(newFavorite))
  }

  render() {
    const {
      logged_in,
      sign_in_path,
      sign_out_path,
      sign_up_path,
      edit_acct_path } = this.props

    const { histories, favorites } = this.state

    return (
      <Router>
        <Header
          logged_in={ logged_in }
          sign_in_path={ sign_in_path }
          sign_out_path={ sign_out_path }
          sign_up_path={ sign_up_path }
          edit_acct_path={ edit_acct_path }
        />

        <Route
        path="/favorite/:id"
        render={ props => <FavoriteShow {...props} favorites={ favorites } />  }
        />
        <Route
          path="/history/:id"
          render={ props => <HistoryShow {...props} addFavorite={this.addFavorite} histories={ histories } />  }
        />
        <Route
          path="/user_history"
          render={ props => <HistoryIndex addFavorite={this.addFavorite} histories={ histories } /> }
        />
        <Route
          path="/user_favorites"
          render={ props => <FavoriteIndex favorites={ favorites } /> }
        />
        <Route path="/about" component={ About } />
        <Route
          // remember to add "exact" for this route or else About page will also appear on the landing page
          exact path="/"
          render={ props => <LandingPage addFavorite={this.addFavorite} logged_in={logged_in} /> }
        />
      </Router>
    )
  }
}

export default App
