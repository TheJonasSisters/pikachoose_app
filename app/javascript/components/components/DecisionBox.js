import React, { Component } from "react";
import { Jumbotron, Button } from "reactstrap";
import Pikachu3 from "./pikachu3.png"

class DecisionBox extends Component {
  constructor() {
    super()
    this.state = {
      filmList: [],
      histories: [],
      favorites: [],
      editable: null
    }
  }

  componentDidMount() {
    this.addHistory()
  }

  // React hook; works like componentDidMount
  // useEffect(() => {
  //   addHistory()
  // })

  addHistory = () => {
    const { filmDecision } = this.props
    let newHistory = {
      film_id: filmDecision.id,
      title: filmDecision.title,
      overview: filmDecision.overview,
      vote_average: filmDecision.vote_average,
      release_date: filmDecision.release_date,
      comment: "Maybe remove comment column from history table?"
    }

    console.log("Added to History:", newHistory)

    fetch("/histories",
    {
      method: 'POST',
      body: JSON.stringify(newHistory),
      headers: {
        'Content-Type': 'application/json'
      }
    })
  }

  addFavorite = (history) => {
    const { filmDecision } = this.props
    console.log("addFavorite(history):",history)
    let newFavorite = {
      film_id: history.id,
      title: history.title,
      overview: history.overview,
      vote_average: history.vote_average,
      release_date: history.release_date,
      comment: "No comment yet"
    }

    // fetch method gets specific history with the id in our back-end and UPDATES it
    fetch("/favorites",
    {
      method: 'POST',
      body: JSON.stringify(newFavorite),
      headers: {
        'Content-Type': 'application/json'
      }
    })

    console.log("newFavorite:", JSON.stringify(newFavorite))
  }

  render() {
    const { filmDecision, logged_in } = this.props
    return (
      <>
        <Jumbotron>
          <p style={{color: "#28A745"}}>Pika-chosen!</p>
          <h4 className="film-title">{filmDecision.title}</h4>
          <p className="my-2">{filmDecision.overview}</p>
          <hr className="my-2" />
          <p>Rating: {filmDecision.vote_average}/10</p>
          <p>Release date: {filmDecision.release_date}</p>
          {logged_in &&
            <p className="lead">
              <Button
                color="success"
                href="/user_favorites"
                onClick={() => this.addFavorite(filmDecision)}>Add to Favorite
              </Button>
            </p>
          }      
        </Jumbotron>
        {/* <img src={Pikachu3} className="pikachu2" /> */}
        <br/>
      </>
    );
  }
};

export default DecisionBox;
