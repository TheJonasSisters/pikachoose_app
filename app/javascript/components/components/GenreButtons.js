import React, { Component } from "react"
import { Button, Container } from "reactstrap"
import Pikachu2 from "./pikachu2.png"

class GenreButtons extends Component {
  constructor(props) {
    super(props)
    this.state = {
      genreList: [],
      selectedGenres: [],
    }
  }

  componentDidMount() {
    this.getGenreList()
  }

  getGenreList = () => {
    const apiKey = process.env.REACT_APP_KEY

    fetch(`https://api.themoviedb.org/3/genre/movie/list?api_key=${apiKey}&language=en-US`)
    .then((response) => {
      if(response.status === 200) {
        return response.json()
      }
    })
    .then((genres) => {
      genres = genres.genres
      this.setState({genreList: genres})
    })
  }

  generateDecision = (e) => {
    // e.persist() is used when you want to access the event properties in an asynchronous way
    // removes the synthetic event from the pool and allow references to the event to be retained by user code.
    e.persist()
    const apiKey = process.env.REACT_APP_KEY
    const { selectedGenres } = this.state
    const { setDisplayToDecisionBox } = this.props

    let genreQuery = selectedGenres.join("&")

    fetch(`https://api.themoviedb.org/3/discover/movie?api_key=${apiKey}&with_genres=${genreQuery}`)
    .then((response) => {
      if(response.status === 200) {
        return response.json()
      }
    })
    .then((films) => {
      let randomNumber = Math.floor(Math.random() * films.results.length);
      let randomFilm = films.results[randomNumber]
      console.log("Generated decision from genre API:", randomFilm);

      if (selectedGenres.length !== 0 && e.target.value === "Submit") {
        // setState of display to DecisionBox
        setDisplayToDecisionBox(randomFilm)
      } else if (e.target.value === "Surprise Me") {
        setDisplayToDecisionBox(randomFilm)
      } else {
        console.log("Something went wrong in generateDecision()")
      }
    })
  }

  activateBtnColor = (button) => {
    // checks to see if the button passed through exists in selectedGenres
    // convert button toString to match datatype in selectedGenres
    if (this.state.selectedGenres.includes(button.toString())) {
      return "#5a6268"
    } else {
      return ""
    }
  }

  userSelectGenre = (e) => {
    e.preventDefault()
    const { selectedGenres } = this.state
    let genre = e.target.value
    // makes a copy to manipulate then setState later
    let userSelection = selectedGenres
    // finds the index of the genre
    const index = userSelection.indexOf(genre);
    // if genre index is -1, push it to the userSelection array
    if (index === -1) {
      userSelection.push(genre);
    }
    // removes the genre from the array if it's clicked on again
    else {
      userSelection.splice(index, 1);
    }

    this.setState({selectedGenres: userSelection})
    console.log("User Selected Genres:", selectedGenres)
  }

  resetUserSelection = (e) => {
    e.preventDefault()
    this.setState({selectedGenres: []})
    // setState is asynchronous so it doesn't necessarily updates instantaniously
    // setTimeout gives setState some time before it logs the state to the console
    setTimeout(() => console.log("User Selection Clears:", this.state.selectedGenres), 1000)
  }

  render() {
    // Genre API keys are "id" and "name"
    const { genreList, selectedGenres } = this.state

    return (
      <Container>
        <div className="genre-grid">
        {genreList.map(genre => {
          return (
            <Button
              key={genre.id}
              value={genre.id}
              className="genre-btn"
              style={{backgroundColor: this.activateBtnColor(genre.id)}}
              onClick={this.userSelectGenre}>{genre.name}
            </Button>
          )
        })}

          <Button
            value="Clear"
            className="genre-btn functional-btn"
            onClick={this.resetUserSelection}>Clear</Button>
          <Button
            value="Submit"
            className="genre-btn functional-btn"
            onClick={this.generateDecision}>Submit</Button>
          <Button
            value="Surprise Me"
            className="genre-btn functional-btn"
            onClick={this.generateDecision}>Surprise Me</Button>
        </div>
        <div>
          <img src={Pikachu2} className="pikachu"/>
          <p>Choose genres (optional).</p>
        </div>
      </Container>
    )
  }
}

export default GenreButtons
