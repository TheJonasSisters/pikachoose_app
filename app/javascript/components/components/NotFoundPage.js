import React from 'react';
import { Link } from 'react-router-dom';
import PageNotFound from './PageNotFound.png';

class NotFoundPage extends React.Component{
    render(){
        return <div>
            <img src={PageNotFound}  />
            <h1> 404 - error </h1>
            <p style={{textAlign:"center"}}>
              <Link to="/">Go to Home </Link>
            </p>
          </div>;
    }
}
export default NotFoundPage;