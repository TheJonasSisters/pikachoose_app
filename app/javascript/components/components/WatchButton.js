import React from "react"
import { Button } from "reactstrap"

const WatchButton = (props) => {
  const { filmDecision, renderDecisionBox } = props

  const addHistory = (event) => {
    event.preventDefault()
    renderDecisionBox()

    let newHistory = {
      film_id: filmDecision.id,
      title: filmDecision.title,
      overview: filmDecision.overview,
      vote_average: filmDecision.vote_average,
      release_date: filmDecision.release_date,
      comment: "Add comment"
    }

    console.log("newHistory:",newHistory)

    fetch("/histories",
    {
      method: 'POST',
      body: JSON.stringify(newHistory),
      headers: {
        'Content-Type': 'application/json'
      }
    }).then((response) => {
      if (response.ok) {
        return response.json()
      }
    })
  }

  return (
    <div id="watch-btn">
      <Button type="button" size="lg" onClick={addHistory}><h5>What to Watch</h5></Button>
    </div>
  )
}

export default WatchButton
