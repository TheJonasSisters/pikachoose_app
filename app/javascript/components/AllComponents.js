import React, { Component } from "react"
import Header from "./components/Header"
import HistoryIndex from "./pages/HistoryIndex.js"
import HistoryShow from "./pages/HistoryShow.js"
import About from "./pages/About.js"
import LandingPage from "./pages/LandingPage.js"
import FavoriteIndex from "./pages/FavoriteIndex.js"
import FavoriteShow from "./pages/FavoriteShow.js"
import NotFoundPage from "./components/NotFoundPage.js"

class AllComponents extends Component {
  
   render(){
        return
   }
}
    
export default AllComponents;

