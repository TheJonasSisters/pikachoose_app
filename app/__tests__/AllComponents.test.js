import React from 'react';;
import Enzyme, {shallow, mount} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import * as components from "../javascript/components/AllComponents";

Enzyme.configure({ adapter: new Adapter() });

Object.keys(components).forEach(DecisionBox=> {
  const Component = components[DecisionBox];
  describe(`Component: ${DecisionBox}`, () => {
    test(`${DecisionBox} renders with default props`, () => {
      const wrapper = shallow(<Component />);
      expect(wrapper).toMatchSnapshot();
    });
  });
}); 

Object.keys(components).forEach(GenreButtons=> {
  const Component = components[GenreButtons];
  describe(`Component: ${GenreButtons}`, () => {
    test(`${GenreButtons} renders with default props`, () => {
      const wrapper = shallow(<Component />);
      expect(wrapper).toMatchSnapshot();
    });
  });
}); 

Object.keys(components).forEach(LandingPage=> {
  const Component = components[LandingPage];
  describe(`Component: ${LandingPage}`, () => {
    test(`${LandingPage} renders with default props`, () => {
      const wrapper = shallow(<LandingPage />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});

Object.keys(components).forEach(FavoriteIndex=> {
  const Component = components[FavoriteIndex];
  describe(`Component: ${FavoriteIndex}`, () => {
    test(`${FavoriteIndex} renders with default props`, () => {
      const wrapper = shallow(<FavoriteIndex />);
      expect(wrapper).toMatchSnapshot();
    });
  });
}); 

Object.keys(components).forEach(Header=> {
  const Component = components[Header];
  describe(`Component: ${Header}`, () => {
    test(`${Header} renders with default props`, () => {
      const wrapper = shallow(<Header />);
      expect(wrapper).toMatchSnapshot();
    });
  });
}); 

Object.keys(components).forEach(About=> {
  const Component = components[About];
  describe(`Component: ${About}`, () => {
    test(`${About} renders with default props`, () => {
      const wrapper = shallow(<About />);
      expect(wrapper).toMatchSnapshot();
    });
  });
}); 

Object.keys(components).forEach(FavoriteShow=> {
  const Component = components[FavoriteShow];
  describe(`Component: ${FavoriteShow}`, () => {
    test(`${FavoriteShow} renders with default props`, () => {
      const wrapper = shallow(<FavoriteShow />);
      expect(wrapper).toMatchSnapshot();
    });
  });
}); 

Object.keys(components).forEach(NotFoundPage=> {
  const Component = components[NotFoundPage];
  describe(`Component: ${NotFoundPage}`, () => {
    test(`${NotFoundPage} renders with default props`, () => {
      const wrapper = shallow(<NotFoundPage />);
      expect(wrapper).toMatchSnapshot();
    });
  });
}); 

Object.keys(components).forEach(HistoryShow=> {
  const Component = components[HistoryShow];
  describe(`Component: ${HistoryShow}`, () => {
    test(`${HistoryShow} renders with default props`, () => {
      const wrapper = shallow(<HistoryShow />);
      expect(wrapper).toMatchSnapshot();
    });
  });
}); 

Object.keys(components).forEach(HistoryIndex=> {
  const Component = components[HistoryIndex];
  describe(`Component: ${HistoryIndex}`, () => {
    test(`${HistoryIndex} renders with default props`, () => {
      const wrapper = shallow(<HistoryIndex />);
      expect(wrapper).toMatchSnapshot();
    });
  });
}); 