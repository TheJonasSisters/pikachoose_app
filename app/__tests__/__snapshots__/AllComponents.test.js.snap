// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: default default renders with default props 1`] = `ShallowWrapper {}`;

exports[`Component: default default renders with default props 2`] = `ShallowWrapper {}`;

exports[`Component: default default renders with default props 3`] = `ShallowWrapper {}`;

exports[`Component: default default renders with default props 4`] = `ShallowWrapper {}`;

exports[`Component: default default renders with default props 5`] = `ShallowWrapper {}`;

exports[`Component: default default renders with default props 6`] = `ShallowWrapper {}`;

exports[`Component: default default renders with default props 7`] = `ShallowWrapper {}`;

exports[`Component: default default renders with default props 8`] = `ShallowWrapper {}`;

exports[`Component: default default renders with default props 9`] = `ShallowWrapper {}`;

exports[`Component: default default renders with default props 10`] = `ShallowWrapper {}`;
